core/controllers/acl_decorators.py

```diff
@@ -45,6 +47,72 @@
 (suggestion_models,) = models.Registry.import_models([models.NAMES.suggestion])
 
 
+def redirect_permanently(handler, redirection_url, expected_return_type):
+    """Redirects to the provided URL if the handler type is not JSON. The
+    status code used would be 301 for permanent redirection.
+
+    Args:
+        handler: function. The function to be decorated.
+        redirection_url: str. The URL to redirect to.
+        expected_return_type: str. The type of the response to be returned
+            in case of errors eg. html, json.
+
+    Raises:
+        PageNotFoundException: The page is not found.
+    """
+    if expected_return_type == feconf.HANDLER_TYPE_JSON:
+        raise handler.PageNotFoundException
+    else:
+        handler.redirect(redirection_url, permanent=True)
+
+
+def validate_and_redirect_based_on_classroom_name(
```

Comment:
This function is doing too many things. Just have it validate; leave the redirection to the decorators. Put the validation function in classroom_services.